libdns.go

```diff
@@ -69,25 +75,61 @@ type RecordSetter interface {
 	// maintain parity with the input. No other records are affected.
 	// It returns the records which were set.
 	//
-	// Records that have an ID associating it with a particular resource
-	// on the provider will be directly replaced. If no ID is given, this
-	// method may use what information is given to do lookups and will
-	// ensure that only necessary changes are made to the zone.
+	// For any (name, type) pair in the input, `SetRecords` ensures that the
+	// only records in the output zone with that (name, type) pair are those
+	// that were provided in the input.
+	//
+	// In RFC 7719 terms, `SetRecords` appends, modifies, or deletes records
+	// in the zone so that for each RRset in the input, the records provided
+	// in the input are the only members of their RRset in the output zone.
 	//
 	// Implementations must honor context cancellation and be safe for
 	// concurrent use.
+	//
+	// Examples:
+	//
+	// 1. Original zone:
+	//    example.com. 3600 IN A   192.0.2.1
+	//    example.com. 3600 IN A   192.0.2.2
+	//    example.com. 3600 IN TXT "hello world"
+	//
+	//    Input:
+	//    example.com. 3600 IN A   192.0.2.3
+	//
+	//    Resultant zone:
+	//    example.com. 3600 IN A   192.0.2.3
+	//    example.com. 3600 IN TXT "hello world"
+	//
+	// 2. Original zone:
+	//    a.example.com. 3600 IN AAAA 2001:db8::1
+	//    a.example.com. 3600 IN AAAA 2001:db8::2
+	//    b.example.com. 3600 IN AAAA 2001:db8::3
+	//    b.example.com. 3600 IN AAAA 2001:db8::4
+	//
+	//    Input:
+	//    a.example.com. 3600 IN AAAA 2001:db8::1
+	//    a.example.com. 3600 IN AAAA 2001:db8::2
+	//    a.example.com. 3600 IN AAAA 2001:db8::5
+	//
+	//    Resultant zone:
+	//    a.example.com. 3600 IN AAAA 2001:db8::1
+	//    a.example.com. 3600 IN AAAA 2001:db8::2
+	//    a.example.com. 3600 IN AAAA 2001:db8::5
+	//    b.example.com. 3600 IN AAAA 2001:db8::3
+	//    b.example.com. 3600 IN AAAA 2001:db8::4
 	SetRecords(ctx context.Context, zone string, recs []Record) ([]Record, error)
 }
 
 // RecordDeleter can delete records from a DNS zone.
 type RecordDeleter interface {
-	// DeleteRecords deletes the given records from the zone if they exist.
-	// It returns the records that were deleted.
+	// DeleteRecords deletes the given records from the zone if they exist in
+	// the zone and exactly match the input. If the input records do not exist
+	// in the zone, they are silently ignored. `DeleteRecords` returns only the
+	// the records that were deleted, and does not return any records that were
+	// provided in the input but did not exist in the zone.
```

Comment:
Rationale: I think that this is the semantics intended by the current interfaces? But it wasn't obvious before if trying to delete a non-existent record would lead to an error or not.

Reply:
Sounds good to me!